In [ ]:
# 1. Write a Python program to print "Hello, World!"
print("Hello, World!")


Hello, World!


In [16]:
# 2. Write a Python program that displays your name and age.
name = "Sahil"
age = 18

print("Name:", name)
print("Age:", age)

Name: Sahil
Age: 18


In [ ]:
# 3. Write code to print all the pre-defined keywords in Python using the keyword library.
import keyword

print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


In [ ]:
# 4. Write a program that checks if a given word is a Python keyword.
import keyword

word = "def"

if keyword.iskeyword(word):
    print(f"{word} is a keyword")
else:
    print(f"{word} is not a keyword")

def is a keyword


In [ ]:
# 5. Create a list and tuple in Python, and demonstrate how attempting to change an element works differently for each.
my_list = [1, 2, 3]
my_tuple = (1, 2, 3)

my_list[0] = 100
print("Modified List:", my_list)

try:
    my_tuple[0] = 100
except TypeError as e:
    print("Error modifying tuple:", e)

Modified List: [100, 2, 3]
Error modifying tuple: 'tuple' object does not support item assignment


In [ ]:
# 6. Write a function to demonstrate the behavior of mutable and immutable arguments.
def modify_data(mutable_list, immutable_int):
    mutable_list.append(4)
    immutable_int = 100
    print("Inside function - List:", mutable_list)
    print("Inside function - Integer:", immutable_int)

my_list = [1, 2, 3]
my_int = 10

modify_data(my_list, my_int)

print("Outside function - List:", my_list)
print("Outside function - Integer:", my_int)

Inside function - List: [1, 2, 3, 4]
Inside function - Integer: 100
Outside function - List: [1, 2, 3, 4]
Outside function - Integer: 10


In [ ]:
# 7. Write a program that performs basic arithmetic operations on two user-input numbers.
num1 = float(input("Enter first number: "))
num2 = float(input("Enter second number: "))

print("Addition:", num1 + num2)
print("Subtraction:", num1 - num2)
print("Multiplication:", num1 * num2)
print("Division:", num1 / num2)

Enter first number: 2.0
Enter second number: 2.5
Addition: 4.5
Subtraction: -0.5
Multiplication: 5.0
Division: 0.8


In [ ]:
# 8. Write a program to demonstrate the use of logical operators.
a = True
b = False

print("a and b:", a and b)
print("a or b:", a or b)
print("not a:", not a)

a and b: False
a or b: True
not a: False


In [ ]:
# 9. Write a Python program to convert user input from string to integer, float, and boolean types.
user_input = input("Enter a value: ")

to_int = int(user_input)
to_float = float(user_input)
to_bool = bool(user_input)

print("Integer:", to_int, type(to_int))
print("Float:", to_float, type(to_float))
print("Boolean:", to_bool, type(to_bool))

Enter a value: 5
Integer: 5 <class 'int'>
Float: 5.0 <class 'float'>
Boolean: True <class 'bool'>


In [ ]:
# 10. Write code to demonstrate type casting with list elements.
str_list = ["1", "2", "3", "4"]
int_list = []

for item in str_list:
    int_list.append(int(item))

print("Original list:", str_list)
print("Type casted list:", int_list)

Original list: ['1', '2', '3', '4']
Type casted list: [1, 2, 3, 4]


In [ ]:
# 11. Write a program that checks if a number is positive, negative, or zero.
number = float(input("Enter a number: "))

if number > 0:
    print("Positive")
elif number < 0:
    print("Negative")
else:
    print("Zero")

Enter a number: -99
Negative


In [ ]:
# 12. Write a for loop to print numbers from 1 to 10.
for i in range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


In [ ]:
# 13. Write a Python program to find the sum of all even numbers between 1 and 50.
total_sum = 0

for i in range(1, 51):
    if i % 2 == 0:
        total_sum += i

print("Sum of even numbers:", total_sum)

Sum of even numbers: 650


In [ ]:
# 14. Write a program to reverse a string using a while loop.
text = "Python"
reversed_text = ""
index = len(text) - 1

while index >= 0:
    reversed_text += text[index]
    index -= 1

print("Original:", text)
print("Reversed:", reversed_text)

Original: Python
Reversed: nohtyP


In [ ]:
# 15. Write a Python program to calculate the factorial of a number provided by the user using a while loop.
num = int(input("Enter a number: "))
factorial = 1
count = num

while count > 0:
    factorial *= count
    count -= 1

print(f"Factorial of {num} is {factorial}")

Enter a number: 5
Factorial of 5 is 120


1. What is Python, and why is it popular?
 --Python is a high-level programming language that is very easy to read and write because its syntax looks a lot like simple English. It is popular because it is versatile; you can use it for almost anything, including web development, data science, artificial intelligence, and automation. It also has a huge community, meaning there are endless free libraries and tools available to help you build things faster.

2. What is an interpreter in Python?  
--An interpreter is like a translator that works line-by-line. Instead of translating your entire code into machine language at once (like a compiler), the Python interpreter reads one line of your code, translates it, executes it, and then moves to the next. This makes it easier to find errors because the program stops exactly where the mistake happened.

3. What are pre-defined keywords in Python?
 --Keywords are special reserved words that Python has already claimed for its own internal use. Each keyword has a specific meaning and triggers a specific action. Examples include if, else, while, for, def, and import.

4. Can keywords be used as variable names?
-- No, you cannot use keywords as variable names. Since Python uses these words to understand the structure of your program (like deciding when to loop or define a function), using them as variable names would confuse the interpreter and cause a syntax error.

5. What is mutability in Python?
 --Mutability refers to the ability to change an object after it has been created. If an object is mutable, you can modify its internal data (like adding items to a list) without creating a completely new object. If it is immutable, you cannot change it once it is made; you would have to create a new one to store a different value.

6. Why are lists mutable, but tuples are immutable?
 --Lists are designed to be dynamic collections where you might need to add, remove, or change items frequently, so they are made mutable. Tuples are designed to be "write-protected" collections for data that shouldn't change, like coordinates or configuration settings. This immutability makes tuples slightly faster and safer for storing constant data.

7. What is the difference between "==" and "is" operators in Python?  
--The == operator checks for equality of value (do these two things look the same?). The is operator checks for identity (are these two things actually the exact same object in memory?).
a == b: True if a and b have the same contents.
a is b: True if a and b point to the same memory address.

8. What are logical operators in Python?  
--Logical operators are used to combine conditional statements.
and: Returns True if both statements are true.
or: Returns True if at least one statement is true.
not: Reverses the result (returns False if the result is true).

9. What is type casting in Python?  
--Type casting is the process of converting a variable from one data type to another. For example, changing the string "123" into the integer 123 so you can perform math on it.

10. What is the difference between implicit and explicit type casting?
--Implicit: Python does it automatically for you. For example, if you add an integer (3) and a float (4.5), Python automatically converts the answer to a float (7.5) to prevent data loss.

--Explicit: You manually force the conversion using functions like int(), str(), or float(). You do this when Python doesn't know how to convert it automatically, like turning a string into a number.

11. What is the purpose of conditional statements in Python?
 --Conditional statements allow your program to make decisions. They let the code execute specific blocks only if certain conditions are met (True). Without them, code would just run straight through from top to bottom every time.

12. How does the elif statement work?  
--elif stands for "else if". It is used to check multiple conditions in a sequence. If the first if condition is False, Python checks the elif. If the elif is True, it runs that block. You can have many elif statements to handle different possibilities one by one.

13. What is the difference between for and while loops?
--For loop: Best used when you know exactly how many times you want to iterate (like looping through a list of 10 items). It runs for a specific set of elements.

--While loop: Best used when you want to loop until a specific condition changes. You don't necessarily know how many times it will run beforehand; it just keeps going as long as the condition is True.

14. Describe a scenario where a while loop is more suitable than a for loop.  
--A while loop is better when you are waiting for an external event or user input. For example, a game menu that keeps asking "Do you want to play again?" requires a while loop because you don't know if the user will say "yes" once or fifty times